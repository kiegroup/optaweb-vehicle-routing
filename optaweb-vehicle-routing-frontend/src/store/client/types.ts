import { Viewport as LeafletViewport } from 'react-leaflet';
import { Action } from 'redux';

export enum ActionType {
  UPDATE_VIEWPORT = 'UPDATE_VIEWPORT',
  RESET_VIEWPORT = 'RESET_VIEWPORT',
}

export interface UpdateViewportAction extends Action<ActionType.UPDATE_VIEWPORT> {
  value: LeafletViewport;
}

export interface ResetViewportAction extends Action<ActionType.RESET_VIEWPORT> {
}

export interface UserViewport {
  isDirty: boolean;
  center: [number, number];
  zoom: number;
}

export type ViewportAction =
  | UpdateViewportAction
  | ResetViewportAction;
