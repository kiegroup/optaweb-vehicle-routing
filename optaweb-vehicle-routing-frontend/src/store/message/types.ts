import { Action } from 'redux';

export enum ActionType {
  RECEIVE_MESSAGE = 'RECEIVE_MESSAGE',
  READ_MESSAGE = 'READ_MESSAGE',
}

export interface ReceiveMessageAction extends Action<ActionType.RECEIVE_MESSAGE> {
  readonly payload: MessagePayload;
}

export interface ReadMessageAction extends Action<ActionType.READ_MESSAGE> {
  readonly id: string;
}

export type MessageAction = ReceiveMessageAction | ReadMessageAction;

export interface MessagePayload {
  readonly id: string;
  readonly text: string;
}

export interface Message extends MessagePayload {
  readonly status: 'new' | 'read';
}
