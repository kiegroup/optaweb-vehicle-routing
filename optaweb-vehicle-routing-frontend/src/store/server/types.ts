import { Action } from 'redux';
import { LatLng } from '../route/types';

export enum ActionType {
  SERVER_INFO = 'SERVER_INFO',
}

export interface ServerInfoAction extends Action<ActionType.SERVER_INFO> {
  value: ServerInfo;
}

export interface Demo {
  name: string;
  visits: number;
}

export type BoundingBox = [LatLng, LatLng];

export interface ServerInfo {
  boundingBox: BoundingBox | null;
  countryCodes: string[];
  demos: Demo[];
}
