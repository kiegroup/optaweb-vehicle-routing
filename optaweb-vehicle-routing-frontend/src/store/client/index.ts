import * as clientOperations from './operations';
import { clientReducer } from './reducers';

export { clientOperations };

export default clientReducer;
