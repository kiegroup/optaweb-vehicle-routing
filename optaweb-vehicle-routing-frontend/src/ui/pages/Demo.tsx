/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  Button,
  ButtonVariant,
  Flex,
  FlexItem,
  FlexModifiers,
  GutterSize,
  InputGroup,
  InputGroupText,
  Split,
  SplitItem,
  Text,
  TextContent,
  TextVariants,
} from '@patternfly/react-core';
import { MinusIcon, PlusIcon } from '@patternfly/react-icons';
import { backendUrl } from 'common';
import * as React from 'react';
import { connect } from 'react-redux';
import { clientOperations } from 'store/client';
import { UserViewport } from 'store/client/types';
import { demoOperations } from 'store/demo';
import { routeOperations, routeSelectors } from 'store/route';
import { LatLng, Location, RouteWithTrack } from 'store/route/types';
import { AppState } from 'store/types';
import { DemoDropdown } from 'ui/components/DemoDropdown';
import LocationList from 'ui/components/LocationList';
import RouteMap from 'ui/components/RouteMap';
import SearchBox, { Result } from 'ui/components/SearchBox';
import { sideBarStyle } from 'ui/pages/common';
import { CapacityInfo, DistanceInfo, VehiclesInfo, VisitsInfo } from 'ui/pages/InfoBlock';

export const ID_CLEAR_BUTTON = 'clear-button';
export const ID_EXPORT_BUTTON = 'export-button';

export interface StateProps {
  distance: string;
  depot: Location | null;
  vehicleCount: number;
  totalCapacity: number;
  totalDemand: number;
  visits: Location[];
  routes: RouteWithTrack[];
  isDemoLoading: boolean;
  boundingBox: [LatLng, LatLng] | null;
  userViewport: UserViewport;
  countryCodeSearchFilter: string[];
  demoNames: string[];
}

export interface DispatchProps {
  loadHandler: typeof demoOperations.requestDemo;
  clearHandler: typeof routeOperations.clearRoute;
  addLocationHandler: typeof routeOperations.addLocation;
  removeLocationHandler: typeof routeOperations.deleteLocation;
  addVehicleHandler: typeof routeOperations.addVehicle;
  removeVehicleHandler: typeof routeOperations.deleteAnyVehicle;
  updateViewport: typeof clientOperations.updateViewport;
}

const mapStateToProps = ({ plan, demo, serverInfo, userViewport }: AppState): StateProps => ({
  distance: plan.distance,
  vehicleCount: plan.vehicles.length,
  totalCapacity: routeSelectors.totalCapacity(plan),
  totalDemand: routeSelectors.totalDemand(plan),
  depot: plan.depot,
  visits: plan.visits,
  routes: plan.routes,
  isDemoLoading: demo.isLoading,
  boundingBox: serverInfo.boundingBox,
  countryCodeSearchFilter: serverInfo.countryCodes,
  // TODO use selector
  // TODO sort demos alphabetically?
  demoNames: (serverInfo.demos && serverInfo.demos.map((value) => value.name)) || [],
  userViewport,
});

const mapDispatchToProps: DispatchProps = {
  loadHandler: demoOperations.requestDemo,
  clearHandler: routeOperations.clearRoute,
  addLocationHandler: routeOperations.addLocation,
  removeLocationHandler: routeOperations.deleteLocation,
  addVehicleHandler: routeOperations.addVehicle,
  removeVehicleHandler: routeOperations.deleteAnyVehicle,
  updateViewport: clientOperations.updateViewport,
};

export type DemoProps = DispatchProps & StateProps;

export interface DemoState {
  selectedId: number;
}

export class Demo extends React.Component<DemoProps, DemoState> {
  constructor(props: DemoProps) {
    super(props);

    this.state = {
      selectedId: NaN,
    };
    this.handleDemoLoadClick = this.handleDemoLoadClick.bind(this);
    this.handleMapClick = this.handleMapClick.bind(this);
    this.handleSearchResultClick = this.handleSearchResultClick.bind(this);
    this.onSelectLocation = this.onSelectLocation.bind(this);
  }

  handleMapClick(e: any) {
    this.props.addLocationHandler({ ...e.latlng, description: '' }); // TODO use reverse geocoding to find address
  }

  handleSearchResultClick(result: Result) {
    this.props.addLocationHandler({ ...result.latLng, description: result.address });
  }

  handleDemoLoadClick(demoName: string) {
    this.props.loadHandler(demoName);
  }

  onSelectLocation(id: number) {
    this.setState({ selectedId: id });
  }

  render() {
    const { selectedId } = this.state;
    const {
      distance,
      depot,
      vehicleCount,
      totalCapacity,
      totalDemand,
      visits,
      routes,
      demoNames,
      isDemoLoading,
      boundingBox,
      userViewport,
      countryCodeSearchFilter,
      addVehicleHandler,
      removeVehicleHandler,
      removeLocationHandler,
      clearHandler,
      updateViewport,
    } = this.props;

    const exportDataSet = () => {
      window.open(`${backendUrl}/api/dataset/export`);
    };

    return (
      // FIXME find a way to avoid these style customizations
      <Split gutter={GutterSize.md} style={{ overflowY: 'auto' }}>
        <SplitItem
          isFilled={false}
          style={sideBarStyle}
        >
          <TextContent>
            <Text component={TextVariants.h1}>Demo</Text>
          </TextContent>
          <SearchBox
            boundingBox={boundingBox}
            countryCodeSearchFilter={countryCodeSearchFilter}
            addHandler={this.handleSearchResultClick}
          />
          <LocationList
            depot={depot}
            visits={visits}
            removeHandler={removeLocationHandler}
            selectHandler={this.onSelectLocation}
          />
        </SplitItem>

        <SplitItem
          isFilled
          style={{ display: 'flex', flexDirection: 'column' }}
        >
          <Flex breakpointMods={[{ modifier: FlexModifiers['justify-content-space-between'] }]}>
            <FlexItem>
              <VisitsInfo visitCount={visits.length} />
            </FlexItem>
            <FlexItem>
              <CapacityInfo totalDemand={totalDemand} totalCapacity={totalCapacity} />
            </FlexItem>
            <FlexItem>
              <Flex>
                <FlexItem>
                  <VehiclesInfo />
                </FlexItem>
                <FlexItem>
                  <InputGroup>
                    <Button
                      variant={ButtonVariant.primary}
                      isDisabled={vehicleCount === 0}
                      onClick={removeVehicleHandler}
                    >
                      <MinusIcon />
                    </Button>
                    <InputGroupText readOnly style={{ minWidth: '2.5em' }}>
                      {vehicleCount}
                    </InputGroupText>
                    <Button
                      variant={ButtonVariant.primary}
                      onClick={addVehicleHandler}
                      data-cy="demo-add-vehicle"
                    >
                      <PlusIcon />
                    </Button>
                  </InputGroup>
                </FlexItem>
              </Flex>
            </FlexItem>
            <FlexItem>
              <DistanceInfo distance={distance} />
            </FlexItem>
            <FlexItem>
              <Button
                id={ID_EXPORT_BUTTON}
                isDisabled={!depot || isDemoLoading}
                style={{ marginBottom: 16, marginLeft: 16 }}
                onClick={exportDataSet}
              >
                Export
              </Button>
              {(depot === null && (
                <DemoDropdown
                  demos={demoNames}
                  onSelect={this.handleDemoLoadClick}
                />
              )) || (
                <Button
                  id={ID_CLEAR_BUTTON}
                  isDisabled={isDemoLoading}
                  style={{ marginBottom: 16, marginLeft: 16 }}
                  onClick={clearHandler}
                  data-cy="demo-clear-button"
                >
                  Clear
                </Button>
              )}
            </FlexItem>
          </Flex>
          <RouteMap
            boundingBox={boundingBox}
            userViewport={userViewport}
            updateViewport={updateViewport}
            selectedId={selectedId}
            clickHandler={this.handleMapClick}
            removeHandler={removeLocationHandler}
            depot={depot}
            routes={routes}
            visits={visits}
          />
        </SplitItem>
      </Split>
    );
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Demo);
