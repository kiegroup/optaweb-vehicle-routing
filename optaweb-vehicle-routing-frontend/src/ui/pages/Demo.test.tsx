/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Button, Dropdown } from '@patternfly/react-core';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { UserViewport } from 'store/client/types';
import { Demo, DemoProps, ID_CLEAR_BUTTON, ID_EXPORT_BUTTON } from './Demo';

describe('Demo page', () => {
  it('should render correctly with no routes', () => {
    const demo = shallow(<Demo {...emptyRouteProps} />);
    expect(toJson(demo)).toMatchSnapshot();
  });

  it('should render correctly with a few routes', () => {
    const demo = shallow(<Demo {...threeLocationsProps} />);
    expect(toJson(demo)).toMatchSnapshot();
  });

  it('clear and export buttons should be disabled when demo is loading', () => {
    const props: DemoProps = {
      ...threeLocationsProps,
      isDemoLoading: true,
    };
    const demo = shallow(<Demo {...props} />);
    expect(toJson(demo)).toMatchSnapshot();

    const clearButton = demo.find(Button).filter(`#${ID_CLEAR_BUTTON}`);
    expect(clearButton.props().isDisabled).toEqual(true);

    clearButton.simulate('click');
    // Doesn't work, probably due to https://github.com/airbnb/enzyme/issues/386
    // expect(props.clearHandler).not.toHaveBeenCalled();

    const exportButton = demo.find(Button).filter(`#${ID_EXPORT_BUTTON}`);
    expect(exportButton.props().isDisabled).toEqual(true);
  });

  it('clear button should replace demo dropdown as soon as there is a depot', () => {
    const props: DemoProps = {
      ...emptyRouteProps,
      depot: {
        id: 1,
        lat: 1,
        lng: 1,
        description: '',
      },
    };
    const demo = shallow(<Demo {...props} />);
    expect(toJson(demo)).toMatchSnapshot();

    const clearButton = demo.find(Button).filter(`#${ID_CLEAR_BUTTON}`);
    expect(clearButton).toHaveLength(1);
    expect(clearButton.props().isDisabled).toEqual(false);

    const exportButton = demo.find(Button).filter(`#${ID_EXPORT_BUTTON}`);
    expect(exportButton).toHaveLength(1);
    expect(exportButton.props().isDisabled).toEqual(false);

    expect(demo.find(Dropdown)).toHaveLength(0);
  });
});

const userViewport: UserViewport = {
  isDirty: false,
  zoom: 1,
  center: [0, 0],
};

const emptyRouteProps: DemoProps = {
  loadHandler: jest.fn(),
  clearHandler: jest.fn(),
  addVehicleHandler: jest.fn,
  removeVehicleHandler: jest.fn,
  addLocationHandler: jest.fn(),
  removeLocationHandler: jest.fn(),
  updateViewport: jest.fn(),

  distance: '0',
  vehicleCount: 0,
  demoNames: ['demo'],
  isDemoLoading: false,
  boundingBox: null,
  userViewport,
  countryCodeSearchFilter: [],

  depot: null,
  routes: [],
  visits: [],
};

const threeLocationsProps: DemoProps = {
  loadHandler: jest.fn(),
  clearHandler: jest.fn(),
  addVehicleHandler: jest.fn,
  removeVehicleHandler: jest.fn,
  addLocationHandler: jest.fn(),
  removeLocationHandler: jest.fn(),
  updateViewport: jest.fn(),

  distance: '10',
  vehicleCount: 8,
  demoNames: ['demo'],
  isDemoLoading: false,
  boundingBox: null,
  userViewport,
  countryCodeSearchFilter: ['XY'],

  depot: {
    id: 1,
    lat: 1.345678,
    lng: 1.345678,
  },

  visits: [{
    id: 2,
    lat: 2.345678,
    lng: 2.345678,
  }, {
    id: 3,
    lat: 3.676111,
    lng: 3.568333,
  }],

  routes: [{
    vehicle: { id: 1, name: 'v1', capacity: 5 },
    visits: [{
      id: 1,
      lat: 1.345678,
      lng: 1.345678,
    }, {
      id: 2,
      lat: 2.345678,
      lng: 2.345678,
    }, {
      id: 3,
      lat: 3.676111,
      lng: 3.568333,
    }],

    track: [],

  }],
};
