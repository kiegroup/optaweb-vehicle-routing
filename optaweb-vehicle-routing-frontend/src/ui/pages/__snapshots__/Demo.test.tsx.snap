// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Demo page clear and export buttons should be disabled when demo is loading 1`] = `
<Split
  className=""
  component="div"
  gutter="md"
  style={
    Object {
      "overflowY": "auto",
    }
  }
>
  <SplitItem
    className=""
    isFilled={false}
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <TextContent>
      <Text
        component="h1"
      >
        Demo
      </Text>
    </TextContent>
    <SearchBox
      addHandler={[Function]}
      boundingBox={null}
      countryCodeSearchFilter={
        Array [
          "XY",
        ]
      }
      searchDelay={500}
    />
    <LocationList
      depot={
        Object {
          "id": 1,
          "lat": 1.345678,
          "lng": 1.345678,
        }
      }
      removeHandler={[MockFunction]}
      selectHandler={[Function]}
      visits={
        Array [
          Object {
            "id": 2,
            "lat": 2.345678,
            "lng": 2.345678,
          },
          Object {
            "id": 3,
            "lat": 3.676111,
            "lng": 3.568333,
          },
        ]
      }
    />
  </SplitItem>
  <SplitItem
    className=""
    isFilled={true}
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <Split
      className=""
      component="div"
      gutter="md"
    >
      <SplitItem
        className=""
        isFilled={true}
      >
        <Grid
          className=""
          gutter={null}
          lg={null}
          md={null}
          sm={null}
          span={null}
          xl={null}
        >
          <GridItem
            className=""
            lg={null}
            lgOffset={null}
            lgRowSpan={null}
            md={null}
            mdOffset={null}
            mdRowSpan={null}
            offset={null}
            rowSpan={null}
            sm={null}
            smOffset={null}
            smRowSpan={null}
            span={8}
            xl={null}
            xlOffset={null}
            xlRowSpan={null}
          >
            <Level
              className=""
              gutter="sm"
            >
              <LevelItem
                style={
                  Object {
                    "display": "flex",
                  }
                }
              >
                <Level
                  className=""
                  gutter="sm"
                >
                  <LevelItem>
                    <InputGroup
                      className=""
                    >
                      <Button
                        isDisabled={false}
                        onClick={[Function]}
                        variant="primary"
                      >
                        <MinusIcon
                          color="currentColor"
                          size="sm"
                          title={null}
                        />
                      </Button>
                      <InputGroupText
                        className=""
                        component="span"
                        readOnly={true}
                      >
                        8
                      </InputGroupText>
                      <Button
                        data-cy="demo-add-vehicle"
                        onClick={[Function]}
                        variant="primary"
                      >
                        <PlusIcon
                          color="currentColor"
                          size="sm"
                          title={null}
                        />
                      </Button>
                    </InputGroup>
                  </LevelItem>
                  <LevelItem>
                    vehicles
                  </LevelItem>
                </Level>
              </LevelItem>
              <LevelItem>
                2 visits
              </LevelItem>
              <LevelItem>
                Total travel time: 10
              </LevelItem>
            </Level>
          </GridItem>
          <GridItem
            className=""
            lg={null}
            lgOffset={null}
            lgRowSpan={null}
            md={null}
            mdOffset={null}
            mdRowSpan={null}
            offset={null}
            rowSpan={null}
            sm={null}
            smOffset={null}
            smRowSpan={null}
            span={4}
            xl={null}
            xlOffset={null}
            xlRowSpan={null}
          />
        </Grid>
      </SplitItem>
      <SplitItem
        className=""
        isFilled={false}
      >
        <Button
          id="export-button"
          isDisabled={true}
          onClick={[Function]}
          style={
            Object {
              "marginBottom": 16,
              "marginLeft": 16,
            }
          }
        >
          Export
        </Button>
        <Button
          data-cy="demo-clear-button"
          id="clear-button"
          isDisabled={true}
          onClick={[MockFunction]}
          style={
            Object {
              "marginBottom": 16,
              "marginLeft": 16,
            }
          }
        >
          Clear
        </Button>
      </SplitItem>
    </Split>
    <RouteMap
      boundingBox={null}
      clickHandler={[Function]}
      depot={
        Object {
          "id": 1,
          "lat": 1.345678,
          "lng": 1.345678,
        }
      }
      removeHandler={[MockFunction]}
      routes={
        Array [
          Object {
            "track": Array [],
            "vehicle": Object {
              "capacity": 5,
              "id": 1,
              "name": "v1",
            },
            "visits": Array [
              Object {
                "id": 1,
                "lat": 1.345678,
                "lng": 1.345678,
              },
              Object {
                "id": 2,
                "lat": 2.345678,
                "lng": 2.345678,
              },
              Object {
                "id": 3,
                "lat": 3.676111,
                "lng": 3.568333,
              },
            ],
          },
        ]
      }
      selectedId={NaN}
      updateViewport={[MockFunction]}
      userViewport={
        Object {
          "center": Array [
            0,
            0,
          ],
          "isDirty": false,
          "zoom": 1,
        }
      }
      visits={
        Array [
          Object {
            "id": 2,
            "lat": 2.345678,
            "lng": 2.345678,
          },
          Object {
            "id": 3,
            "lat": 3.676111,
            "lng": 3.568333,
          },
        ]
      }
    />
  </SplitItem>
</Split>
`;

exports[`Demo page clear button should replace demo dropdown as soon as there is a depot 1`] = `
<Split
  className=""
  component="div"
  gutter="md"
  style={
    Object {
      "overflowY": "auto",
    }
  }
>
  <SplitItem
    className=""
    isFilled={false}
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <TextContent>
      <Text
        component="h1"
      >
        Demo
      </Text>
    </TextContent>
    <SearchBox
      addHandler={[Function]}
      boundingBox={null}
      countryCodeSearchFilter={Array []}
      searchDelay={500}
    />
    <LocationList
      depot={
        Object {
          "description": "",
          "id": 1,
          "lat": 1,
          "lng": 1,
        }
      }
      removeHandler={[MockFunction]}
      selectHandler={[Function]}
      visits={Array []}
    />
  </SplitItem>
  <SplitItem
    className=""
    isFilled={true}
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <Split
      className=""
      component="div"
      gutter="md"
    >
      <SplitItem
        className=""
        isFilled={true}
      >
        <Grid
          className=""
          gutter={null}
          lg={null}
          md={null}
          sm={null}
          span={null}
          xl={null}
        >
          <GridItem
            className=""
            lg={null}
            lgOffset={null}
            lgRowSpan={null}
            md={null}
            mdOffset={null}
            mdRowSpan={null}
            offset={null}
            rowSpan={null}
            sm={null}
            smOffset={null}
            smRowSpan={null}
            span={8}
            xl={null}
            xlOffset={null}
            xlRowSpan={null}
          >
            <Level
              className=""
              gutter="sm"
            >
              <LevelItem
                style={
                  Object {
                    "display": "flex",
                  }
                }
              >
                <Level
                  className=""
                  gutter="sm"
                >
                  <LevelItem>
                    <InputGroup
                      className=""
                    >
                      <Button
                        isDisabled={true}
                        onClick={[Function]}
                        variant="primary"
                      >
                        <MinusIcon
                          color="currentColor"
                          size="sm"
                          title={null}
                        />
                      </Button>
                      <InputGroupText
                        className=""
                        component="span"
                        readOnly={true}
                      >
                        0
                      </InputGroupText>
                      <Button
                        data-cy="demo-add-vehicle"
                        onClick={[Function]}
                        variant="primary"
                      >
                        <PlusIcon
                          color="currentColor"
                          size="sm"
                          title={null}
                        />
                      </Button>
                    </InputGroup>
                  </LevelItem>
                  <LevelItem>
                    vehicles
                  </LevelItem>
                </Level>
              </LevelItem>
              <LevelItem>
                0 visits
              </LevelItem>
              <LevelItem>
                Total travel time: 0
              </LevelItem>
            </Level>
          </GridItem>
          <GridItem
            className=""
            lg={null}
            lgOffset={null}
            lgRowSpan={null}
            md={null}
            mdOffset={null}
            mdRowSpan={null}
            offset={null}
            rowSpan={null}
            sm={null}
            smOffset={null}
            smRowSpan={null}
            span={4}
            xl={null}
            xlOffset={null}
            xlRowSpan={null}
          />
        </Grid>
      </SplitItem>
      <SplitItem
        className=""
        isFilled={false}
      >
        <Button
          id="export-button"
          isDisabled={false}
          onClick={[Function]}
          style={
            Object {
              "marginBottom": 16,
              "marginLeft": 16,
            }
          }
        >
          Export
        </Button>
        <Button
          data-cy="demo-clear-button"
          id="clear-button"
          isDisabled={false}
          onClick={[MockFunction]}
          style={
            Object {
              "marginBottom": 16,
              "marginLeft": 16,
            }
          }
        >
          Clear
        </Button>
      </SplitItem>
    </Split>
    <RouteMap
      boundingBox={null}
      clickHandler={[Function]}
      depot={
        Object {
          "description": "",
          "id": 1,
          "lat": 1,
          "lng": 1,
        }
      }
      removeHandler={[MockFunction]}
      routes={Array []}
      selectedId={NaN}
      updateViewport={[MockFunction]}
      userViewport={
        Object {
          "center": Array [
            0,
            0,
          ],
          "isDirty": false,
          "zoom": 1,
        }
      }
      visits={Array []}
    />
  </SplitItem>
</Split>
`;

exports[`Demo page should render correctly with a few routes 1`] = `
<Split
  className=""
  component="div"
  gutter="md"
  style={
    Object {
      "overflowY": "auto",
    }
  }
>
  <SplitItem
    className=""
    isFilled={false}
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <TextContent>
      <Text
        component="h1"
      >
        Demo
      </Text>
    </TextContent>
    <SearchBox
      addHandler={[Function]}
      boundingBox={null}
      countryCodeSearchFilter={
        Array [
          "XY",
        ]
      }
      searchDelay={500}
    />
    <LocationList
      depot={
        Object {
          "id": 1,
          "lat": 1.345678,
          "lng": 1.345678,
        }
      }
      removeHandler={[MockFunction]}
      selectHandler={[Function]}
      visits={
        Array [
          Object {
            "id": 2,
            "lat": 2.345678,
            "lng": 2.345678,
          },
          Object {
            "id": 3,
            "lat": 3.676111,
            "lng": 3.568333,
          },
        ]
      }
    />
  </SplitItem>
  <SplitItem
    className=""
    isFilled={true}
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <Split
      className=""
      component="div"
      gutter="md"
    >
      <SplitItem
        className=""
        isFilled={true}
      >
        <Grid
          className=""
          gutter={null}
          lg={null}
          md={null}
          sm={null}
          span={null}
          xl={null}
        >
          <GridItem
            className=""
            lg={null}
            lgOffset={null}
            lgRowSpan={null}
            md={null}
            mdOffset={null}
            mdRowSpan={null}
            offset={null}
            rowSpan={null}
            sm={null}
            smOffset={null}
            smRowSpan={null}
            span={8}
            xl={null}
            xlOffset={null}
            xlRowSpan={null}
          >
            <Level
              className=""
              gutter="sm"
            >
              <LevelItem
                style={
                  Object {
                    "display": "flex",
                  }
                }
              >
                <Level
                  className=""
                  gutter="sm"
                >
                  <LevelItem>
                    <InputGroup
                      className=""
                    >
                      <Button
                        isDisabled={false}
                        onClick={[Function]}
                        variant="primary"
                      >
                        <MinusIcon
                          color="currentColor"
                          size="sm"
                          title={null}
                        />
                      </Button>
                      <InputGroupText
                        className=""
                        component="span"
                        readOnly={true}
                      >
                        8
                      </InputGroupText>
                      <Button
                        data-cy="demo-add-vehicle"
                        onClick={[Function]}
                        variant="primary"
                      >
                        <PlusIcon
                          color="currentColor"
                          size="sm"
                          title={null}
                        />
                      </Button>
                    </InputGroup>
                  </LevelItem>
                  <LevelItem>
                    vehicles
                  </LevelItem>
                </Level>
              </LevelItem>
              <LevelItem>
                2 visits
              </LevelItem>
              <LevelItem>
                Total travel time: 10
              </LevelItem>
            </Level>
          </GridItem>
          <GridItem
            className=""
            lg={null}
            lgOffset={null}
            lgRowSpan={null}
            md={null}
            mdOffset={null}
            mdRowSpan={null}
            offset={null}
            rowSpan={null}
            sm={null}
            smOffset={null}
            smRowSpan={null}
            span={4}
            xl={null}
            xlOffset={null}
            xlRowSpan={null}
          />
        </Grid>
      </SplitItem>
      <SplitItem
        className=""
        isFilled={false}
      >
        <Button
          id="export-button"
          isDisabled={false}
          onClick={[Function]}
          style={
            Object {
              "marginBottom": 16,
              "marginLeft": 16,
            }
          }
        >
          Export
        </Button>
        <Button
          data-cy="demo-clear-button"
          id="clear-button"
          isDisabled={false}
          onClick={[MockFunction]}
          style={
            Object {
              "marginBottom": 16,
              "marginLeft": 16,
            }
          }
        >
          Clear
        </Button>
      </SplitItem>
    </Split>
    <RouteMap
      boundingBox={null}
      clickHandler={[Function]}
      depot={
        Object {
          "id": 1,
          "lat": 1.345678,
          "lng": 1.345678,
        }
      }
      removeHandler={[MockFunction]}
      routes={
        Array [
          Object {
            "track": Array [],
            "vehicle": Object {
              "capacity": 5,
              "id": 1,
              "name": "v1",
            },
            "visits": Array [
              Object {
                "id": 1,
                "lat": 1.345678,
                "lng": 1.345678,
              },
              Object {
                "id": 2,
                "lat": 2.345678,
                "lng": 2.345678,
              },
              Object {
                "id": 3,
                "lat": 3.676111,
                "lng": 3.568333,
              },
            ],
          },
        ]
      }
      selectedId={NaN}
      updateViewport={[MockFunction]}
      userViewport={
        Object {
          "center": Array [
            0,
            0,
          ],
          "isDirty": false,
          "zoom": 1,
        }
      }
      visits={
        Array [
          Object {
            "id": 2,
            "lat": 2.345678,
            "lng": 2.345678,
          },
          Object {
            "id": 3,
            "lat": 3.676111,
            "lng": 3.568333,
          },
        ]
      }
    />
  </SplitItem>
</Split>
`;

exports[`Demo page should render correctly with no routes 1`] = `
<Split
  className=""
  component="div"
  gutter="md"
  style={
    Object {
      "overflowY": "auto",
    }
  }
>
  <SplitItem
    className=""
    isFilled={false}
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <TextContent>
      <Text
        component="h1"
      >
        Demo
      </Text>
    </TextContent>
    <SearchBox
      addHandler={[Function]}
      boundingBox={null}
      countryCodeSearchFilter={Array []}
      searchDelay={500}
    />
    <LocationList
      depot={null}
      removeHandler={[MockFunction]}
      selectHandler={[Function]}
      visits={Array []}
    />
  </SplitItem>
  <SplitItem
    className=""
    isFilled={true}
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
      }
    }
  >
    <Split
      className=""
      component="div"
      gutter="md"
    >
      <SplitItem
        className=""
        isFilled={true}
      >
        <Grid
          className=""
          gutter={null}
          lg={null}
          md={null}
          sm={null}
          span={null}
          xl={null}
        >
          <GridItem
            className=""
            lg={null}
            lgOffset={null}
            lgRowSpan={null}
            md={null}
            mdOffset={null}
            mdRowSpan={null}
            offset={null}
            rowSpan={null}
            sm={null}
            smOffset={null}
            smRowSpan={null}
            span={8}
            xl={null}
            xlOffset={null}
            xlRowSpan={null}
          >
            <Level
              className=""
              gutter="sm"
            >
              <LevelItem
                style={
                  Object {
                    "display": "flex",
                  }
                }
              >
                <Level
                  className=""
                  gutter="sm"
                >
                  <LevelItem>
                    <InputGroup
                      className=""
                    >
                      <Button
                        isDisabled={true}
                        onClick={[Function]}
                        variant="primary"
                      >
                        <MinusIcon
                          color="currentColor"
                          size="sm"
                          title={null}
                        />
                      </Button>
                      <InputGroupText
                        className=""
                        component="span"
                        readOnly={true}
                      >
                        0
                      </InputGroupText>
                      <Button
                        data-cy="demo-add-vehicle"
                        onClick={[Function]}
                        variant="primary"
                      >
                        <PlusIcon
                          color="currentColor"
                          size="sm"
                          title={null}
                        />
                      </Button>
                    </InputGroup>
                  </LevelItem>
                  <LevelItem>
                    vehicles
                  </LevelItem>
                </Level>
              </LevelItem>
              <LevelItem>
                0 visits
              </LevelItem>
              <LevelItem>
                Total travel time: 0
              </LevelItem>
            </Level>
          </GridItem>
          <GridItem
            className=""
            lg={null}
            lgOffset={null}
            lgRowSpan={null}
            md={null}
            mdOffset={null}
            mdRowSpan={null}
            offset={null}
            rowSpan={null}
            sm={null}
            smOffset={null}
            smRowSpan={null}
            span={4}
            xl={null}
            xlOffset={null}
            xlRowSpan={null}
          />
        </Grid>
      </SplitItem>
      <SplitItem
        className=""
        isFilled={false}
      >
        <Button
          id="export-button"
          isDisabled={true}
          onClick={[Function]}
          style={
            Object {
              "marginBottom": 16,
              "marginLeft": 16,
            }
          }
        >
          Export
        </Button>
        <DemoDropdown
          demos={
            Array [
              "demo",
            ]
          }
          onSelect={[Function]}
        />
      </SplitItem>
    </Split>
    <RouteMap
      boundingBox={null}
      clickHandler={[Function]}
      depot={null}
      removeHandler={[MockFunction]}
      routes={Array []}
      selectedId={NaN}
      updateViewport={[MockFunction]}
      userViewport={
        Object {
          "center": Array [
            0,
            0,
          ],
          "isDirty": false,
          "zoom": 1,
        }
      }
      visits={Array []}
    />
  </SplitItem>
</Split>
`;
